dask/array/core.py

```diff
@@ -3007,6 +3008,24 @@ def ensure_int(f):
     return i
 
 
+@functools.lru_cache
```

Comment:
I bet you could use this function internally if you wrote a caching decorator that cached both by `id` and `hash`. That way you check `id` first, and then with the `hash` if that fails.

Reply:
Oh this is interesting, I'll take a look at this but will merge that one for now